Add Js_exn.anyToExnInternal function

ryyppy opened the pull request:
I am trying to express an idiomatic `Promise.catch` approach, which usually receives an error like this:

```
type t<+'a> = Js.Promise.t<'a>
type error = Js.Promise.error

@bs.send
external catch: (t<'a>, @bs.uncurry (error => 'b)) => t<'b> = "catch"
```

Now, `error`'s value is determined by following scenarios (in a scope of a `Promise.then`):
- 1) `raise(MyOwnError)`
- 2) `Promise.reject(MyOwnError)`
- 3) `Js.Exn.raiseError`
- 4) `someJsFunctionThatThrowsError()`
- 5) `someJsFnThatThrowsNonError()`

Case 1) and 2) are fine, because they are already making sure that our `error` is pattern matchable.

Generally I would have liked to just type the catch value to be of `type exn` and just use the explicit exception patterns and a default case for the other values, like this:

```rescript
@val external asyncParseFail: unit => Promise.t<string> = "asyncParseFail"
    asyncParseFail()->catch((e:exn) => {
      switch e {
      | err =>
        switch Js.Exn.asJsExn(e) {
          | Some(v) => Js.log(v)
          | None => Js.log("not a JS Exn")
        }
        Js.Exn.asJsExn(err)->Belt.Option.flatMap(v => Js.Exn.message(v)) ==
          Some("Unexpected token . in JSON at position 1")
      }
    })
  }
```

In this particular example we have a problem though: Since `asyncParseFail` throws an `Error` object, it will not respond to the `asJsExn` convertion, because it is not wrapped in a `Js.Exn.Error` value, therefore I have no chance of seeing if it is a JS Error or not.

Which means we need to rely on `Js.Exn.isCamlExceptionOrVariant` and if so, try to coerce it somehow to the appropriate value:

```
// Promise.res
exception Unknown(Js.Exn.t)
let handleError = (e: error) => {
   if unsafeToExn(e)->Js.Exn.isCamlExceptionOrOpenVariant { 
   unsafeToExn(e) 
   } else { 
   Unknown(unsafeToJsExn(e)) 
   } 
}
```

With this function in place, we will have an easier time "lifting" the passed values to a fine granular type that allows more logical access to the data:

```
    asyncParseFail()->catch(e => {
      let success = switch handleError(e) {
      | UnknownError(obj) =>
        switch Js.Exn.message(obj) {
        | Some(msg) => Js.log("It's a js exn: " ++ msg)
        | None => Js.log("it's not")
        }
        true
      | err => false
      }
```

Now that `UnknownError` exception is extremely annoying, because we have a standardized value that is `Js.Exn.Error(Js.Exn.t)` already.


So the idea of this PR is to add an (internal) function `Js.Exn.unsafeAnyToExn` function, that takes any value, and lifts it to a `Js.Exn.Error` value, in case it is not already a exn / open variant:

```res
let unsafeAnyToExn: ('a) => exn
``` 

With this in place we can simplify our `handleError` to the following:

```
// let handleError: error => exn
let handleError = e => {
  Js.Exn.unsafeAnyToExn(e)
}
```

Now in our program, instead of using `Unknown`, we are using our `Js.Exn.Error` value (like in normal try / catch scenarios):

```diff
    asyncParseFail()->catch(e => {
      let success = switch handleError(e) {
+      | Js.Exn.Error(obj) =>
        switch Js.Exn.message(obj) {
        | Some(msg) => Js.log("It's a js exn: " ++ msg)
        | None => Js.log("it's not")
        }
        true
      | err => false
      }
```

This would help us implement a cleaner Promise error handling implementation without adding too much overhead, while being familiar with exception handling!

bobzhang commented on jscomp/others/js_exn.ml:
```diff
@@ -45,6 +45,12 @@ external makeError : string -> error = "Error" [@@bs.new]
+  if isCamlExceptionOrOpenVariant obj then
+    (Obj.magic obj : exn)
+  else
+    Error ((Obj.magic obj) : t)
```
think twice, this is a safe function. but it is encouraged only for internal usage, maybe we can rename it into `anyToExnInternal`

ryyppy replied:
what is considered to be internal? do you mean i should create some function in `caml_js_exceptions`, and use an external to bind to it? I browsed the code yesterday, but didn't really understand how one defines a "real internal" function.

> maybe we can rename it into anyToExnInternal

Yes, it's not really unsafe, so I'd like this name change

bobzhang commented:
I share same ideas. I suggest we hide handError inside the binding of `catch` so that user only need write:
```
   asyncParseFail()->catch(e => { // here `e` is already wrapped as exn
      let success = switch (e) { 
       | Js.Exn.Error(obj) =>
        switch Js.Exn.message(obj) {
        | Some(msg) => Js.log("It's a js exn: " ++ msg)
        | None => Js.log("it's not")
        }
        true
      | err => false
      }
```

ryyppy commented:
> I share same ideas. I suggest we hide handError inside the binding of catch

I like this idea. Catch is most likely not going to be used very often, so the extra JS code is probably negligible. Will adapt the promise library to address that change!

ryyppy commented:
@bobzhang okay, changed the name of the function and added an appropriate note in the docheader that this function is internal (which means that it might be changed / removed in the future). Since this is supposed to be internal API, I feel comfortable proposing this change to allow us a nicer design for our promise bindings.

ryyppy commented:
Just adapted the promise catching logic, this is how it could potentially look like then:

```rescript
let asyncParseFail: unit => Js.Promise.t<string> = %raw(`
  function() {
    return new Promise((resolve) => {
      var result = JSON.parse("{..");
      return resolve(result);
    })
  }
  `)

open Promise
let _ =
  asyncParseFail()
  ->catch((e) => {
    let s = switch e {
    | TestError("some rejected value") => "caught Test error"
    | Js.Exn.Error(obj) =>
      "Some JS error: " ++ Js.Exn.message(obj)->Belt.Option.getWithDefault("")
    | _ => "some other error"
    }
    resolve(s)
  })
  ->then(msg => {
    Js.log(msg)
  })

```

Way better, since we can now reuse the `Js.Exn.Error` again, and `catch` is now handling an `exn` type instead of an `error`.

bobzhang commented:
For the record, there's a duplication between Js.Exn.anyToExnInternal and Caml_js_exceptions.internalToOCamlException, we may provide such as a primitive in the future
